### ISSUE TITLE
Extract lambda refactoring

### ISSUE THREAD

**Issue by dantleech** [phpactor/phpactor#1192]
Given a bunch of code

```
$bar = foobar();
$baz = $foo->findBar($bar->getBlah());
if (!$baz) throw new Exception($bar->getBlah());
```

Extract lambda:

```
$baz = function (Blah $nlahr) {
    $baz = $foo->findBAr($blah);
    if (!$baz) throw new Exception($blah);
    return $baz;
}, $bar->getBlah());
```

or similar :thinking:

**Comment by BladeMF** [comment]
That would have similar code to `extract method` but generate different code, right?

**Comment by dantleech** [comment]
Yeah, I guess it's just a variant on extract method.

**Comment by BladeMF** [comment]
Yes, because you still need to determine what the arguments are. I am not sure how you decide between `use` and an argument...

**Comment by camilledejoye** [comment]
I would put everything as parameter, I don't like `use`: it breaks things in my mind.

**Comment by dantleech** [comment]
For me one use case for lambdas is removing temporary variables - in which case the otherwise temporary variables would be parameters, while everything else would be `use`. But anyway, removing temporary vars is probably another refactoring.

**Comment by camilledejoye** [comment]
What do you mean by temporary vars ?

**Comment by dantleech** [comment]
```
$foo = $this->foobar()->barfoo();
if (false === $this->doFoo($foo)) {
    throw new \Exception('Foo is not a foo' . $foo);
}
return $foo;
```
has the temporary var `$foo` which polutes the scope, so:

```
return (function ($foo) {
    if (false === $this->doFoo($foo)) {
    }
}

**Comment by camilledejoye** [comment]
OK but then I'm not sure we can differentiate temporary variable from not temporary.  
Maybe a variable is not temporary if it uses after the refactored code.

But then why pass them in the `use` more than as argument ?

**Comment by dantleech** [comment]
arguably if the parent scope already has the necessary vars, `use ($foo, $bar)` can be far easier to read than parsing the function to the end and looking for the arguments, but deciding isn't trivial and would probably change from case-to-case anyway. maybe support both :shrug:  in anycase `use` might be easier.
